### ISSUE TITLE
Compiler Inconsistency: Can compile structs with function in global scripts but not in room scripts

### ISSUE THREAD

**Issue by edmundito** [adventuregamestudio/ags#187]
It turns out that in a global script/script module source you can declare a struct with methods in them such as:

``` C++
// Module.asc
struct MyStruct {
  import function MyFunction();
};

function MyStruct::MyFunction() {
}
```

Trying to do the same thing in a room script will cause a script error because the compiler cannot properly find MyFunction:

> Variable 'MyStruct::MyFunction' is already imported

**Comment by ivan-mogilko** [comment]
Confirmed to still present in 3.5.0.16.

**Comment by ivan-mogilko** [comment]
Should have looked into this earlier. The reason is not the compiler itself, but that room scripts are compiled with an additional option:

https://github.com/adventuregamestudio/ags/blob/master/Editor/AGS.Native/ScriptCompiler.cpp#L76

**Comment by ivan-mogilko** [comment]
I sort of *think* this might have been to prevent occasional mistakes in the times when there were no script modules but only a global script and global header. Then this could have been a solution for users mistakenly placing imported function implementation in the room script. But that's just a guess.

I think we may remove this, because it makes no sense now.

**Comment by ivan-mogilko** [comment]
Fixed by e08de235c7154a82c681709aa69208a3a52972b8

**Comment by fernewelten** [comment]
As far as I understand the issue with "overriding imports":

Let's do variables first and then functions.

In general, you have a module body `B.asc` and the corresponding module header `B.ash`: 
* When the body of the module is compiled, then the header file and the body file are compiled together.
* When all the other entities are compiled, then only the header file is included in front of the entities.

In the former case, the header file will have lots of `import` variables, but the engine should not import the variables really: To the contrary, the body file will provide local definitions for them, and the engine should work with these local versions. 

However, those variables need to be EXported in this case so that they can be available to other modules. So whenever a body file provides local definitions for a variable that has been `import`ed, then this local definition will “override” the `import` statement of the header file and convert it into an `export` instead.

Now, rooms are a special case. The AGS tries to economize on memory usage, and so it compiles each room into a separate “room program”. Whenever the player character changes rooms at runtime, the program for the old room is swapped out and the program for the new room is swapped in. This means that all the variables in room files are __not__ available to __other__ room files.

The room programs are assembled by joining (amongst others): 
* `globalscript.ash` (the header)
*  `globalscript.asc` (the body)
* and the respective room script.

All the variables that `globalscript.asc` has (the __body!!__ file) will be automatically available in room scripts because this body file is included in all the room programs. So there's no need to define those variables as `import` in the `globalscript.ash` header file and override them as exports later on. OTOH, all the variables in a room script __may not__ be passed to other room programs, so they may not be imported in the header file and may not later be overridden as exports. 

In practice, this means that the `globalscript.ash` header is limited to just `struct` and `enum` definitions and `#define`s. `globalscript.ash` shouldn't contain any `import` statements, but for technical reasons the compiler makes sure instead that no `import`s are overridden.

Thus a compiler flag that forbids "overriding imports" which is set whenever room scripts are compiled and unset everywhere else.

That's not the doing of the new compiler; the old compiler had it this way already. Both the compilers work with a special a flag. When it is set, they refuse overrides of `import` statements. -- --


__As concerns functions,__ there's a special flag that makes the compiler export all the functions automatically when set. This flag is __always__ set. 

When you put a struct into `globalscript.ash` (the header) that contains a function, it is obligatory that this function must be `import`ed. Thus when you define the function with body later on, this will trigger the “No overwriting imports” error. As far as I know, one way of fixing this is putting the function body into `globalscript.asc` (the body) using the `int func(this struct, …)` syntax. 

A more general fix for the issue would be to compile `globalscript.ash` / `globalscript.asc` only once, separately, and link it to the respective room scripts. But this is dicey. We'd need to think through what the ramifications of that would be. The room scripts would only be able to reference what's put into `globalscript.ash`, the header file; whatever only is in `globalscript.asc` would be out of scope. This may break most existing programs.
